### PR TITLE
CSV export for transactions

### LINKED ISSUE #366: Allow the data from any /*.all endpoint to be exported to CSV

### PR DESCRIPTION
**Pull request by T-Dnzt**
Issue/Task Number: #366 
closes #366 

# Overview

This PR adds CSV export capabilities for transactions to the eWallet. Exports work a bit differently depending on the storage backend being used:

- S3: The file will be loaded from DB, 500 records at a time, and streamed through a pipeline that will turn it into CSV and send each chunk to a S3 bucket file.
- GCS: The file will first be written locally via streaming before being uploaded directly to Google Cloud Storage.
- Local: The file will be written locally via streaming (500 records loaded in memory at a time).

# Changes

- Added a `CSVExporter` module that can be used to generate an export.
- 4 endpoints were added to deal with those exports: 
  - `/export.all`: get all the exports the current user/key has generated
  - `/export.get`: get a specific report, generating a signed URL to download (when using GCS or S3)
  - `/export.download`: download a specific export when using local storage.

# Implementation Details

GCS didn't support proper streaming for files, or at least, I wasn't able to find a good library/documentation on how to do it in a timely manner - so instead, it's writing the file to the local storage first, making it complicated to use with providers like Heroku. For those, it'll make more sense to use S3 since nothing gets written to the filesystem, instead chunk are sent automatically to S3 once they reach a certain size.

There is currently no websocket available, it should come in a future PR.

### CODE REVIEW

**Comment by mederic-p** [comment]
I cloned and tried with the open api spec to generate a transaction report with the following params but I'm getting a 500 error:

File: apps/ewallet_api/config/config.exs [hunk 1]
```diff
@@ -67,7 +67,8 @@ config :ewallet_api, :api_versions, %{
 # what format to respond with.
 # Run `mix deps.clean --build mime` when updaing this mapping.
 config :mime, :types, %{
-  "application/vnd.omisego.v1+json" => ["json"]
+  "application/vnd.omisego.v1+json" => ["json"],
+  "text/vnd.omisego.v1+csv" => ["csv"]
```
Comment (unnawut) [inline apps/ewallet_api/config/config.exs]: 👍

File: apps/ewallet_api/mix.exs [hunk 2]
```diff
@@ -11,7 +11,7 @@ defmodule EWalletAPI.Mixfile do
       lockfile: "../../mix.lock",
       elixir: "~> 1.4",
       elixirc_paths: elixirc_paths(Mix.env),
-      elixirc_options: [warnings_as_errors: Mix.env != :test],
+      elixirc_options: [warnings_as_errors: false],
```
Comment (unnawut) [inline apps/ewallet_api/mix.exs]: Can remove the whole line

File: apps/ewallet_db/lib/ewallet_db/export.ex [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  Retrieves an export using one or more fields.
+  """
+  @spec get_by(map() | keyword(), keyword()) :: %Export{} | nil | no_return()
+  def get_by(fields, opts \\ []) do
+    Export
+    |> Repo.get_by(fields)
+    |> preload_option(opts)
+  end
+
+  def init(export, schema, count, estimated_size, originator) do
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/export.ex]: Can you add a function doc for this one? I'm not sure when to use `Export.init/5` vs. `Export.insert/1`.

File: apps/ewallet_db/lib/ewallet_db/export.ex [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  """
+  @spec insert(map()) :: {:ok, %Export{}} | {:error, Ecto.Changeset.t()}
+  def insert(attrs) do
+    %Export{}
+    |> create_changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/export.ex]: 💥

File: apps/ewallet_db/lib/ewallet_db/export.ex [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      adapter: Config.get(:file_storage_adapter),
+      schema: schema,
+      total_count: count,
+      estimated_size: estimated_size,
+      originator: originator
+    })
+  end
+
+  @doc """
+
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/export.ex]: 💥

File: apps/ewallet_db/lib/ewallet_db/uploaders/file.ex [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  @acl :private
+  @versions [:original]
+
+  def validate({file, _}) do
+    ~w(.csv) |> Enum.member?(Path.extname(file.file_name))
+  end
+
+  # Override the storage directory:
+  def storage_dir(_, _) do
+    "private/uploads/#{Mix.env()}/exports"
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/uploaders/file.ex]: @sirn Would this work with distillery?

File: apps/ewallet_db/mix.exs [hunk 7]
```diff
@@ -11,7 +11,7 @@ defmodule EWalletDB.Mixfile do
       lockfile: "../../mix.lock",
       elixir: "~> 1.4",
       elixirc_paths: elixirc_paths(Mix.env),
-      elixirc_options: [warnings_as_errors: Mix.env != :test],
+      elixirc_options: [warnings_as_errors: false],
```
Comment (unnawut) [inline apps/ewallet_db/mix.exs]: Can remove the whole line

File: apps/local_ledger/mix.exs [hunk 8]
```diff
@@ -10,7 +10,7 @@ defmodule LocalLedger.Mixfile do
       deps_path: "../../deps",
       lockfile: "../../mix.lock",
       elixir: "~> 1.5",
-      elixirc_options: [warnings_as_errors: Mix.env != :test],
+      elixirc_options: [warnings_as_errors: false],
```
Comment (unnawut) [inline apps/local_ledger/mix.exs]: Can remove the entire line

File: apps/load_tester/README.md [hunk 9]
```diff
@@ -12,3 +12,30 @@ Chaperon.run_load_test(LoadTester.Runner); nil
 - Output folder resides in the sub-app, not at the umbrella app
 - Easier way to inspect the latest session state
 - More readable way to process the result rather than `:with_result` option
+
+## Usage
```
Comment (unnawut) [inline apps/load_tester/README.md]: Thanks! 🙏

File: apps/local_ledger_db/mix.exs [hunk 10]
```diff
@@ -11,7 +11,7 @@ defmodule LocalLedgerDB.Mixfile do
       lockfile: "../../mix.lock",
       elixir: "~> 1.5",
       elixirc_paths: elixirc_paths(Mix.env),
-      elixirc_options: [warnings_as_errors: Mix.env != :test],
+      elixirc_options: [warnings_as_errors: false],
```
Comment (unnawut) [inline apps/local_ledger_db/mix.exs]: Can remove the entire line

File: apps/url_dispatcher/mix.exs [hunk 11]
```diff
@@ -10,7 +10,7 @@ defmodule UrlDispatcher.Mixfile do
       deps_path: "../../deps",
       lockfile: "../../mix.lock",
       elixir: "~> 1.5",
-      elixirc_options: [warnings_as_errors: Mix.env != :test],
+      elixirc_options: [warnings_as_errors: false],
```
Comment (unnawut) [inline apps/url_dispatcher/mix.exs]: Can remove the entire line

File: apps/ewallet/lib/ewallet/exporters/csv_exporter.ex [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+             ],
+             name: {:global, export.uuid}
+           ),
+         {:ok, export} <-
+           Export.update(export, %{
+             pid: PidHelper.pid_to_binary(pid),
+             originator: %Exporter{}
+           }),
+         :ok <- GenServer.cast(pid, :upload) do
+      {:ok, pid, export}
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/exporters/csv_exporter.ex]: What's the pid used for?

Comment (T-Dnzt) [reply apps/ewallet/lib/ewallet/exporters/csv_exporter.ex]: The PID is used to identify the process responsible for exporting. It is returned to the client and could potentially be used to check what's up with a process. Right now, it's mostly used in order to have functional tests :D

File: apps/ewallet_db/lib/ewallet_db/uploaders/file.ex (hunk 6 above)
Comment (sirn) [reply apps/ewallet_db/lib/ewallet_db/uploaders/file.ex]: Neeed to change to `Application.get_env(:ewallet, :env)`. Mix is not present in Distillery.

File: apps/admin_api/lib/admin_api/v1/controllers/export_controller.ex [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    end
+  end
+
+  def get(conn, _), do: handle_error(conn, :invalid_parameter)
+
+  def download(conn, %{"id" => id}) do
+    with %Export{} = export <- Export.get(id) || {:error, :unauthorized},
+         :ok <- permit(:get, conn.assigns, export),
+         path <- Path.join(Application.get_env(:ewallet, :root), export.path),
+         true <- File.exists?(path) || {:error, :file_not_found} do
```
[the remark this answers is not in the document]
Comment (T-Dnzt) [reply apps/admin_api/lib/admin_api/v1/controllers/export_controller.ex]: Or a new custom error: `export_not_local` maybe?

File: apps/ewallet/lib/ewallet/adapters/gcs_adapter.ex [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        handle_successful_upload(args.export, path)
+
+      {:error, error} ->
+        {:ok, export} = AdapterHelper.store_error(args.export, error)
+        {:error, export}
+    end
+  end
+
+  defp handle_successful_upload(export, path) do
+    {:ok, export} = AdapterHelper.update_export(export, Export.completed(), 100)
```
[the remark this answers is not in the document]
Comment (unnawut) [reply apps/ewallet/lib/ewallet/adapters/gcs_adapter.ex]: I just made this more explicit so `AdapterHelper.update_export/3` doesn't assume that pid is to be nullified if not provided.

Removing the pid requires `AdapterHelper.update_export/4` with 4th (pid) argument as nil.

File: apps/ewallet/lib/ewallet/adapters/gcs_adapter.ex [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    File.rm(path)
```
[the remark this answers is not in the document]
Comment (unnawut) [reply apps/ewallet/lib/ewallet/adapters/gcs_adapter.ex]: Sorry I wasn't clear. Successful uploads delete the local file, but failed uploads won't at the moment. I just updated the code to delete the file on failures as well.

**Comment by unnawut** [comment]
> I cloned and tried with the open api spec to generate a transaction report with the following params but I'm getting a 500 error:
> 
> Params:
> 
> ```
> {
>   "match_all": [{"field":"created_at", "value": "2018-12-20T10:25:21.155165", "comparator": "gt"}],
>   "match_any": [],
>   "sort_by": "created_at",
>   "sort_dir": "desc"
> }
> ```
> Error:
> 
> ```
> [info] Sent 500 in 15ms
> [error] #PID<0.1651.0> running UrlDispatcher.Plug (cowboy_protocol) terminated
> Server: localhost:4000 (http)
> Request: POST /api/admin/transaction.export
> ** (exit) an exception was raised:
>     ** (Postgrex.Error) ERROR 42P10 (invalid_column_reference): for SELECT DISTINCT, ORDER BY expressions must appear in select list
>         (ecto) lib/ecto/adapters/sql.ex:436: Ecto.Adapters.SQL.execute_and_cache/7
>         (ecto) lib/ecto/repo/queryable.ex:130: Ecto.Repo.Queryable.execute/5
>         (ecto) lib/ecto/repo/queryable.ex:35: Ecto.Repo.Queryable.all/4
>         (ecto) lib/ecto/repo/queryable.ex:76: Ecto.Repo.Queryable.one!/4
>         (ewallet) lib/ewallet/exporters/csv_exporter.ex:13: EWallet.CSVExporter.start/4
>         (ewallet) lib/ewallet/gates/export_gate.ex:23: EWallet.ExportGate.export/4
>         (admin_api) lib/admin_api/v1/controllers/transaction_controller.ex:34: AdminAPI.V1.TransactionController.export/2
>         (admin_api) lib/admin_api/v1/controllers/transaction_controller.ex:1: AdminAPI.V1.TransactionController.action/2
>         (admin_api) lib/admin_api/v1/controllers/transaction_controller.ex:1: AdminAPI.V1.TransactionController.phoenix_controller_pipeline/2
>         (admin_api) lib/admin_api/endpoint.ex:1: AdminAPI.Endpoint.instrument/4
>         (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
>         (phoenix) lib/phoenix/router/route.ex:147: Phoenix.Router.Route.forward/4
>         (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
>         (admin_api) lib/admin_api/endpoint.ex:1: AdminAPI.Endpoint.plug_builder_call/2
>         (admin_api) lib/admin_api/endpoint.ex:3: AdminAPI.Endpoint."call (overridable 2)"/2
>         (admin_api) lib/plug/debugger.ex:122: AdminAPI.Endpoint."call (overridable 3)"/2
>         (admin_api) lib/admin_api/endpoint.ex:1: AdminAPI.Endpoint."call (overridable 4)"/2
>         (admin_api) lib/admin_api/endpoint.ex:1: AdminAPI.Endpoint.call/2
>         (plug_cowboy) lib/plug/cowboy/handler.ex:18: Plug.Adapters.Cowboy.Handler.upgrade/4
>         (cowboy) /Users/mederic/work/elixir/ewallet/deps/cowboy/src/cowboy_protocol.erl:442: :cowboy_protocol.execute/4
> ```
> Curl:
> 
> ```
> curl -X POST "http://localhost:4000/api/admin/transaction.export" -H "accept: application/vnd.omisego.v1+json" -H "Authorization: TO REPLACE" -H "Content-Type: application/vnd.omisego.v1+json" -d "{\"match_all\":[{\"field\":\"created_at\",\"value\":\"2018-12-20T10:25:21.155165\",\"comparator\":\"gt\"}],\"match_any\":[],\"sort_by\":\"created_at\",\"sort_dir\":\"desc\"}"
> ```

Looks like Ecto isn't able to build an aggregation query with both distinct and order_by.

Ecto tried to query with this:

```sql
SELECT DISTINCT "uuid" FROM "transaction"
--WHERE ...
  ORDER BY "inserted_at" DESC
```

instead of

```sql
SELECT DISTINCT "uuid", "inserted_at" FROM "transaction"
--WHERE ...
  ORDER BY "inserted_at" DESC
```

And hence the error.

Fixed in 131fa2e.

**Pull request merged**